### DISCUSSION TITLE
Support for create widget

### DISCUSSION THREAD

**Discussion by conor-mac-aoidh** [ArtifexSoftware/mupdf.js#124] [Ideas]
I would like to create form fields (a.k.a. Widgets) with muPDF in JS. Looking at the API docs ([`PDFPage.createAnnotation`](https://mupdf.readthedocs.io/en/latest/mutool-run-js-api.html#createAnnotation)), this does not appear to be possible. However, it is possible in the C library and in PyMuPDF.

What would be involved in adding support for this in mupdf.js ?

**Comment by jamie-lemon** [comment]
@conor-mac-aoidh I think this is an oversight in the docs - you should be able to do: `page.createAnnotation("Widget")`

API for a `PDFWidget` is here: https://mupdfjs.readthedocs.io/en/latest/classes/PDFWidget.html , and it extends `PDFAnnotation` - https://mupdfjs.readthedocs.io/en/latest/classes/PDFAnnotation.html

Thanks for noticing this - we will update the docs.

**Comment by conor-mac-aoidh** [comment]
Thank you for clarifying.

I also see a `addAnnotation` method [here](https://mupdfjs.readthedocs.io/en/latest/classes/PDFPage.html#addAnnotation). This does not appear to be defined in the lastest version of mupdf.js.

I was able to create a widget with `createAnnotation`. This actually returns a `PDFAnnotation` though and doesn't have the widget methods such as `setTextValue`. I was able to get the new widget for now using `page.getWidgets`.

Some additional questions.
- I can get the field type of the widget using `widget.getFieldType`, but how do I set the field type? I would like to create a checkbox widget.
- I ran the example `annotations-test`. It draws a text widget. The widget is not interactive - there is no field to input a value. Is this something that needs to be implemented on top of mupdf.js?
- In `annotations-test`, a widget with value "3" is drawn. But if I swap it for a PDF with acroforms, then no text values are displayed / widgets found. I can see the values in other PDF viewers:
![image](https://github.com/user-attachments/assets/14be2286-5574-4f87-9e52-fc509b445b37)
<img width="1093" alt="image" src="https://github.com/user-attachments/assets/97fd796f-bb73-4b59-8184-07583f80d481" />

**Comment by jamie-lemon** [comment]
Thanks for this info. I've been looking more into our state-of-play here a bit more.

The `addAnnotation` method is in the top level wrapper file `mupdfjs.js`, this wraps `mupdf.js` and aims to provide a more user-friendly and convenient API. ( i.e. This file: https://github.com/ArtifexSoftware/mupdf.js/blob/master/src/mupdfjs.ts ) . It is recommended to use MuPDF.js at this level as it aims to bring more tested & complete functionality to the project (e.g. in the future we really need to do `addWidget()` as you will see... ), but of course it is up to you what level you want to operate at.

Widget support is very much work in progress and from looking at line 3587 here: https://github.com/ArtifexSoftware/mupdf.js/blob/master/src/mupdf.ts I can see there are a few TODOs around interactivity which need to be resolved. Right now I don't believe you can set a field type. So although we have some form of Widget support I would say it is very much read only.

By the way, I'm confused when you say that the annotations-test example creates a widget, all I see in that example is the _reading_ widgets info from a loaded PDF. Can you clarify?

In summary we want to reach this kind of benchmark in the short term:

https://pymupdf.readthedocs.io/en/latest/widget.html#supported-widget-types

However, at present, I'm realising that we are probably right at present to state that Widget _creation_ is not supported in MuPDF.js (because it is not fully functional) and the current documentation is in this sense "correct". There is much more to do here so I would recommend watching the repo for further developments and definitely to [find MuPDF.js on Discord ](https://discord.gg/zpyAHM7XtF) to keep in touch with what is going on ( and to nudge us :) ). 

In terms of the ongoing development, the approach here will be to:

- complete the functionality in the main mupdf file (`mupdf.ts`) 
- then add a new convenience method for `addWidget` to the top level wrapper (`mupdfjs.ts`) 

In this way I hope a usable interface can be presented for widgets.

**Comment by conor-mac-aoidh** [comment]
Thanks for your detailed reply.

I will have a go at using the top-level API. Also I'll join the discord when I return in January :)

> By the way, I'm confused when you say that the annotations-test example creates a widget, all I see in that example is the reading widgets info from a loaded PDF. Can you clarify?

Apologies for the confusion, I meant that the annotation-test renders the text widget with the value "3", but it cannot be edited.

**Reply by jamie-lemon** [reply]
Have a good break!
